### ISSUE TITLE
Determine whether OpenJDK respects method redefinitions on intrinsically inlined methods

### ISSUE THREAD

**Issue by fjeremic** [eclipse/openj9#2381]
Some methods in the JCL are replaced by hand-written intrinsics to improve performance. Such methods are tagged in JDK9+ in OpenJDK with an `@HotSpotIntrinsicCandidate` annotation signaling the reader that the corresponding method is treated specially by the VM.

An example of such a method which is critical for String performance can be found in [1]. For OpenJ9 to recognize this method and replace it by an equivalent IL sequence we need to know whether we must respect method definitions on these intrinsic candidates. To avoid having JCM inconsistencies we must determine what OpenJDK does when such a method is redefined.

The task here is to determine whether OpenJDK respects method redefinitions on methods annotated with `@HotSpotIntrinsicCandidate` annotation.

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk10/blob/e180ae0ddb9c92f4ed88d0e82246a65242ad5bcb/src/java.base/share/classes/java/lang/StringUTF16.java#L146-L154

**Comment by fjeremic** [comment]
To determine what happens on OpenJDK vs. OpenJ9 we perform the following experiment:

1. Create a new class in `java.lang` which is able to call the `java.lang.StringUTF16.toBytes` method:

```java
package java.lang;

public final class StringRedefine {
    public static char[] staticValue;

    public static byte[] callHotSpotIntrinsicCandidate()
    {
        return StringUTF16.toBytes(staticValue, 0, staticValue.length);
    }
}
```

2. Modify java/lang/StringUTF16 to introduce two static fields `a` and `b` and inside the `toBytes` method increment field `a` upon every call:

```java
    public static int a = 0;
    public static int b = 0;

    @HotSpotIntrinsicCandidate
    public static byte[] toBytes(char[] value, int off, int len) {
        byte[] val = newBytesFor(len);
        for (int i = 0; i < len; i++) {
            putChar(val, i, value[off]);
            off++;
        }
        a++;
        return val;
    }
```

3. Create a test class which exercises our custom `java.lang` class we created in step 1:

```java
public class Test
{
    public static byte[] escapingField;

    public static void main(String[] args) throws InterruptedException
    {
        StringRedefine.staticValue = new char[] { '1', '2', '3', '4', '5', '6', '7', '8' };

        for (int i = 0; i < Integer.parseInt(args[0]); ++i)
        {
            foo();
        }
        System.out.println("StringUTF16.a = " + StringUTF16.a);
        System.out.println("StringUTF16.b = " + StringUTF16.b);
    }

    public static void foo()
    {
        escapingField = StringRedefine.callHotSpotIntrinsicCandidate();
    }
}
```

Note that the number of times we will execute `java.lang.StringUTF16.toBytes` is deterministic depending on the command line input by the user.

4. Create a JVMTI agent using javassist (https://github.com/jboss-javassist/javassist) which will redefine the StringUTF16 class and redefine the `toBytes` method to increment `b` instead of `a`. The way we determine which bytecode to edit is to simply print the value of `bytes` in the below snippet, then recompile StringUTF16 with incrementing `b` instead of `a` and print the `bytes` in the below snippet again. Then we diff the output and find which indicies in the array need to be modified. This is where the magic numbers 2610, 3018, 16109, 16113 come from:

```java
import java.lang.instrument.*;
import javassist.ByteArrayClassPath;
import javassist.ClassPool;
import javassist.CtClass;
import javassist.CtConstructor;
import java.security.ProtectionDomain;
import java.lang.*;
import java.util.*;

public class RedefinitionAgent
{
    public static Instrumentation instrument;
    public static void premain(String agentArgument, Instrumentation inst) {
        instrument = inst;
        System.err.println("Redefinition Agent started ...");
        System.out.println("Redefinition Agent started ...");
        Thread thread = new Thread("Weapon Thread") {
            public void run() {
                try {
                System.err.println("Weapon thread sleeping for 20 sec ...");
                System.out.println("Weapon thread sleeping for 20 sec ...");
                Thread.sleep(20000);
                System.err.println("Weapon thread launching payload ...");
                System.out.println("Weapon thread launching payload ...");

                ClassPool cp = ClassPool.getDefault();
                CtClass cc = cp.get("java.lang.StringUTF16");
                byte[] bytes = cc.toBytecode();
                System.out.println("Changing bytecodes to now increment StringUTF16.b");
                bytes[2610] = 110;
                bytes[3018] = 108;
                bytes[16109] = 100;
                bytes[16113] = 31;
                cc.detach();
                instrument.redefineClasses(new ClassDefinition(java.lang.StringUTF16.class, bytes));
                System.err.println("StringUTF16 redefined ... weapon terminated");
                System.out.println("StringUTF16 redefined ... weapon terminated");
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        };
        thread.start();
    }
}
```

Using the above agent will redefine the StringUTF16 class 20 seconds into the run. We will use this to determine what happens in OpenJDK vs. OpenJ9.

5. Put it all together:

```
# To compile StringRedefine and the new StringUTF16
sdk-openj9/bin/javac --patch-module java.base=. ./patched-java-lang/StringRedefine.java ./patched-java-lang/StringUTF16.java -d patched-java-lang

# To compile Test
sdk-openj9/bin/javac --patch-module java.base=./patched-java-lang Test.java

# To compile the agent
../sdk-openj9/bin/javac -cp ./javassist.jar --patch-module java.base=./patched-java-lang RedefinitionAgent.java;../sdk-openj9/bin/jar -cvfm RedefinitionAgent.jar Redefinition.mf javassist.jar RedefinitionAgent*;
```

6. Now we run with the JIT enabled and disabled using OpenJDK and OpenJ9:

```
# OpenJDK and OpenJ9 builds used:
# https://github.com/AdoptOpenJDK/openjdk10-nightly/releases/download/jdk-10%2B23-201807111552/OpenJDK10_s390x_Linux_201807111552.tar.gz
# https://github.com/AdoptOpenJDK/openjdk10-openj9-nightly/releases/download/jdk-10.0.1%2B10-201807091639/OpenJDK10-OPENJ9_s390x_Linux_201807091639.tar.gz

# OpenJDK with JIT enabled
> sdk-openjdk/bin/java --patch-module java.base=patched-java-lang -javaagent:./redefinition-agent/RedefinitionAgent.jar Test 600000000
Redefinition Agent started ...
Redefinition Agent started ...
Weapon thread sleeping for 20 sec ...
Weapon thread sleeping for 20 sec ...
Weapon thread launching payload ...
Weapon thread launching payload ...
Changing bytecodes to now increment StringUTF16.b
StringUTF16 redefined ... weapon terminated
StringUTF16 redefined ... weapon terminated
StringUTF16.a = 131902
StringUTF16.b = 0

# OpenJ9 with JIT enabled
> sdk-openj9/bin/java --patch-module java.base=patched-java-lang -javaagent:./redefinition-agent/RedefinitionAgent.jar Test 600000000
Redefinition Agent started ...
Redefinition Agent started ...
Weapon thread sleeping for 20 sec ...
Weapon thread sleeping for 20 sec ...
Weapon thread launching payload ...
Weapon thread launching payload ...
Changing bytecodes to now increment StringUTF16.b
StringUTF16 redefined ... weapon terminated
StringUTF16 redefined ... weapon terminated
StringUTF16.a = 254650038
StringUTF16.b = 345349962

# OpenJDK with JIT disabled (interpreter only mode)
> sdk-openjdk/bin/java --patch-module java.base=patched-java-lang -javaagent:./redefinition-agent/RedefinitionAgent.jar -Xint Test 900000
Redefinition Agent started ...
Redefinition Agent started ...
Weapon thread sleeping for 20 sec ...
Weapon thread sleeping for 20 sec ...
Weapon thread launching payload ...
Weapon thread launching payload ...
Changing bytecodes to now increment StringUTF16.b
StringUTF16 redefined ... weapon terminated
StringUTF16 redefined ... weapon terminated
StringUTF16.a = 295374
StringUTF16.b = 604626

# OpenJ9 with JIT disabled (interpreter only mode)
> sdk-openj9/bin/java --patch-module java.base=patched-java-lang -javaagent:./redefinition-agent/RedefinitionAgent.jar -Xint Test 9000000
Redefinition Agent started ...
Redefinition Agent started ...
Weapon thread sleeping for 20 sec ...
Weapon thread sleeping for 20 sec ...
Weapon thread launching payload ...
Weapon thread launching payload ...
Changing bytecodes to now increment StringUTF16.b
StringUTF16 redefined ... weapon terminated
StringUTF16 redefined ... weapon terminated
StringUTF16.a = 3164792
StringUTF16.b = 5835208
```

Attaching a minimal reproduction setup here:
[redefine-StringUTF16.zip](https://github.com/eclipse/openj9/files/2190102/redefine-StringUTF16.zip)

**Comment by fjeremic** [comment]
From step 6 above we can see that OpenJDK in fact does **not** respect method redefinitions on methods annotated with `@HotSpotIntrinsicCandidate`. From the OpenJDK JIT run we can see that early on in the run we compile `StringRedefine.callHotSpotIntrinsicCandidate` which will inline the intrinsic for `java.lang.StringUTF16.toBytes` inside its compiled body. This explains why `StringUTF16.a` stops getting incremented after some iterations (as soon as `StringRedefine.callHotSpotIntrinsicCandidate` gets JIT compiled in fact).

Now when the RedefinitionAgent fires after 20 seconds one would expect `StringUTF16.b` to have a non-zero value as we're starting to increment this other static variable now. However we can see from the OpenJDK JIT results this is not the case. `StringUTF16.b == 0`  in the run, meaning we never respected the method redefinition.

Contrasting this to when we disable JIT in OpenJDK and what happens in OpenJ9 we see the expected behavior of `StringUTF16.b` having a non-zero value.

In conclusion: OpenJDK does **not** respect method redefinitions on intrinsically inlined methods and hence OpenJ9 is free to inline such methods without HCR guards as well.
